src/mlpack/core/optimizers/cne/cne.hpp

```diff
@@ -0,0 +1,220 @@
+ *
+ * The whole process then repeats for multiple generation until atleast one of 
+ * the termination criteria is met-
+ *
+ * 1) The final value of the objective function (Not considered if not provided)
+ * 2) The maximum number of generation reached (optional but highly recommended)
+ * 3) Minimum change in best fitness values between two consecutive generations 
+ *    should be greater than a threshold value (Not considered if not provided) 
+ *
+ * After which the final value and the parameters are returned by the optimize
+ * method.
+ *
+ * For CNE to work, a DecomposableFunctionType template parameter is required.
+ * This class must implement the following function:
+ *
+ * size_t NumFunctions();
+ * 
+ * NumFunctions() should return the number of functions or data points in
+ * the dataset.
+ *
+ * double Evaluate(const arma::mat& coordinates, const size_t i);
+ *
+ * The parameter i refers to which individual function is being evaluated.
+ * The function will evaluate the objective function on the data point i in
+ * the dataset.
+ */
+class CNE
+{
+ public:
+/**
+ * Constructor for the CNE optimizer.
+ * 
+ * The default values provided over here are not necessarily suitable for a
+ * given function. Therefore it is highly recommended to adjust the
+ * parameters according to the problem.
+ *
+ * @param populationSize The number of candidates in the population
+ * @param maxGenerations The maximum number of generations allowed for CNE
+ * @param mutationProb   Probability that a weight will get mutated 
+ * @param mutationSize   The range of mutation noise to be added. This range
+ *                       is between 0 and mutationSize
+ * @param selectPercent  The percentage of candidates to select to become the
+ *                       the next generation
+ * @param finalValue     The final value of the objective function for  
+ *                       termination. Not considered if not provided.
+ * @param fitnessHist    Minimum change in best fitness values between two
+ *                       consecutive generations should be greater than
+ *                       threshold. Not considered if not provided by user.
+ */
+CNE(const size_t populationSize = 15,
+    const size_t maxGenerations = 30,
+    const double mutationProb = 0.5,
+    const double mutationSize = 0.5,
+    const double selectPercent = 0.4,
+    const double finalValue = DBL_MAX,
+    const double fitnessHist = DBL_MAX);
+
+/**
+ * Optimize the given function using CNE. The given
+ * starting point will be modified to store the finishing point of the
+ * algorithm, and the final objective value is returned.
+ *
+ * @tparam DecomposableFunctionType Type of the function to be optimized.
+ * @param function Function to optimize.
+ * @param iterate Starting point (will be modified).
+ * @return Objective value of the final point.
+ */
+  template<typename DecomposableFunctionType>
+  double Optimize(DecomposableFunctionType& function, arma::mat& answer);
```

Comment:
Should we use ``iterate`` instead of ``answer`` as parameter name, to be consistent with the rest of the codebase?